### ISSUE TITLE
Discrepancy in orbital overlap integral for f+3 and f-3 basis functions between PySCF and ORCA

### ISSUE THREAD

**Issue by anmorgunov** [pyscf/pyscf#2352]
Hello,

Is there any reason why I should expect a sign discrepancy between values in PySCF (2.6.2) and ORCA (6.0.0) of overlap integral of f+3 and f-3 functions (in def2-tzvp and/or def2-tzvpp basis)?

pyscf overlap matrix is obtained as follows:

```python
mol = pyscf.gto.Mole()
mol.atom = """
O 	0.0000 	0.0000 	0.1173
H 	0.0000 	0.7572 	-0.4692
H 	0.0000 	-0.7572 	-0.4692
"""
mol.basis = "def2-tzvp"
mol.build()
pyscf_S = mol.intor("int1e_ovlp")
```

orca overlap matrix is obtained by running a SP calculation with the same geometry & basis, parsing the result, and permuting some rows & cols (ORCA lists orbitals in a slightly different order, e.g. dz^2 dxz dyz dx2-y2 dxy instead of dxy dyz dz2 dxz dx2-y2 as in pyscf)

If needed, I attach S matrices from orca/pyscf as txt files 

[orca_S.txt](https://github.com/user-attachments/files/16489005/orca_S.txt)
[pyscf_labels.txt](https://github.com/user-attachments/files/16489006/pyscf_labels.txt)
[pyscf_S.txt](https://github.com/user-attachments/files/16489007/pyscf_S.txt)

the pyscf and orca S matrices differ in the sign of integrals involving f+3 and f-3 functions:

![overlap_diff](https://github.com/user-attachments/assets/5a469936-3c50-42f7-84a8-339c1dca69f6)

you can reproduce the figure as:

```python
import plotly.graph_objs as go
import numpy as np

with open("pyscf_labels.txt") as f:
    pyscf_labels = f.readlines()

orca_pyscf_format = np.loadtxt("orca_S.txt")
pyscf_S = np.loadtxt("pyscf_S.txt")

fig = go.Figure(
    go.Heatmap(
        z=pyscf_S - orca_pyscf_format,
        x=pyscf_labels,
        y=pyscf_labels,
        colorscale="RdBu",
        zmax=0.5,
        zmin=-0.5,
    )
)

fig.show()
```

Multiplying the entries corresponding to integrals involving f+3 and f-3 orbitals by -1 results in identical overlap matrices.

Is there a reason for this sign discrepancy?

**Comment by susilehtola** [comment]
> Is there a reason for this sign discrepancy?

Most definitely yes. Quantum chemistry programs are famously non-interoperable, since most developers only work within a single code. As a result, Gaussian basis set technology comes with many, many conventions for ordering of both cartesian and spherical functions, function norms, as well as phases; this is nicely illustrated [by a xkcd comic](https://xkcd.com/927/)

Frank Neese actually broached this topic in an email exchange with myself in November 2023:
> Another issue with libraries would be our - let me phrase this carefully - idiosyncratic (to avoid the word idiotic; ok, that wasn’t careful :-)) choices of basis function phases that does not follow any of the accepted conventions. It is deeply engraved in ORCA and it can't really be changed. We understand the phase relationships to standard definitions of real spherical harmonics but have no good explanations of why ours are the way they are. It is and will be documented in detail in the upcoming manual that describes the JSON interface between ORCA and the outer world that will be designed to have a two way communication with outside programs.

I thus suggest you have a look at the Orca 6.0.0 manual to see for any other differences.

**Comment by anmorgunov** [comment]
> Quantum chemistry programs are famously non-interoperable

I know it's been true in the past, but I thought things have became much better in recent 5-10 years? We can even see it in this example -- values of overlap integrals match (within 1e-12 at least) for all basis functions except for f+3 and f-3, and even there it's only a change of sign, the values are effectively identical. The same is true for a more complicated system with ~1k basis functions (where I found the discrepancy to begin with). Which is why I thought it should be isolated to some peculiarity in which either orca or pyscf defines the f±3 orbitals.

and I think this quote:

> choices of basis function phases that does not follow any of the accepted conventions

could be the explanation. So thank you very much for your response! I

**Comment by tberkel** [comment]
Indeed, the behavior you see just means that those AO basis functions are
defined with opposite signs in the code. There is no unique sign for the
definition of atomic orbitals, only the possibility for conventions.

**Comment by h-larsson** [comment]
ORCA's convention is mentioned in https://www.faccts.de/docs/orca/6.0/manual/contents/detailed/orca_2json.html

**Comment by susilehtola** [comment]
> > Quantum chemistry programs are famously non-interoperable
> 
> I know it's been true in the past, but I thought things have become much better in recent 5-10 years? We can even see it in this example -- values of overlap integrals match (within 1e-12 at least) for all basis functions except for f+3 and f-3, and even there it's only a change of sign, the values are effectively identical. The same is true for a more complicated system with ~1k basis functions (where I found the discrepancy to begin with). Which is why I thought it should be isolated to some peculiarity in which either orca or pyscf defines the f±3 orbitals.

This is a completely different thing. The programs can perform the same calculations, and if the basis set used is the same (there *can be* some major differences in basis sets between programs!), you can expect to reach sub-$\mu E_h$ agreement between different programs in Hartree-Fock and post-Hartree-Fock calculations, if you converge the calculations with respect to all numerical parameters e.g. integral screening and SCF convergence threshold.

Interoperability means being able to use the programs together. As you have already observed in your comparison, there are significant differences between the data formats employed by the programs, even though they are doing the same thing. These incompatibilities mean that you can't really use two programs together. What you have done is a workaround to this incompatibility for the combination ORCA+PySCF, yet since we have $\mathcal{O}(40)$ Gaussian quantum chemistry packages, it remains extremely laborious to try to figure out how to get every program to exchange data with each other.

What we really need is a common standard for exchanging data. Current workarounds include using Gaussian's formatted checkpoint format, which was only designed for use with Gaussian, or the Molden visualization program's limited format  which is actually not a standard at all since different programs use different formats for this file!

[TrexIO](https://github.com/trex-coe/trexio) is the closest attempt I know of to get everyone on the same page with respect to storing basis sets, wave functions, and integrals. Others include MolSSI's [QCEngine](https://github.com/MolSSI/QCEngine) and [MDI](https://github.com/MolSSI-MDI/MDI_Library), but they are limited to passing data between codes.

FWIW we will be organizing a WATOC satellite conference in Helsinki next summer after WATOC, where I hope to have in-depth discussions on this topic, for example.

**Comment by 1234zou** [comment]
If you go to higher angular momenta, different signs become more and more. Here are [some codes](https://gitlab.com/jxzou/mokit/-/blob/master/src/read_mkl.f90?ref_type=heads#L1064) from [MOKIT](https://gitlab.com/jxzou/mokit)
```
 forall(i = 1:nfmark:1)
  coeff(f_mark(i)+5,:) = -coeff(f_mark(i)+5,:)
  coeff(f_mark(i)+6,:) = -coeff(f_mark(i)+6,:)
 end forall

 forall(i = 1:ngmark:1)
  coeff(g_mark(i)+5,:) = -coeff(g_mark(i)+5,:)
  coeff(g_mark(i)+6,:) = -coeff(g_mark(i)+6,:)
  coeff(g_mark(i)+7,:) = -coeff(g_mark(i)+7,:)
  coeff(g_mark(i)+8,:) = -coeff(g_mark(i)+8,:)
 end forall

 forall(i = 1:nhmark:1)
  coeff(h_mark(i)+5,:) = -coeff(h_mark(i)+5,:)
  coeff(h_mark(i)+6,:) = -coeff(h_mark(i)+6,:)
  coeff(h_mark(i)+7,:) = -coeff(h_mark(i)+7,:)
  coeff(h_mark(i)+8,:) = -coeff(h_mark(i)+8,:)
 end forall

 forall(i = 1:nimark:1)
  coeff(i_mark(i)+5,:) = -coeff(i_mark(i)+5,:)
  coeff(i_mark(i)+6,:) = -coeff(i_mark(i)+6,:)
  coeff(i_mark(i)+7,:) = -coeff(i_mark(i)+7,:)
  coeff(i_mark(i)+8,:) = -coeff(i_mark(i)+8,:)
 end forall
```

If you want some conversion between PySCF <-> ORCA, the [py2orca](https://gitlab.com/jxzou/qcinstall/-/blob/main/%E5%88%A9%E7%94%A8MOKIT%E4%BB%8EPySCF%E5%90%91%E5%85%B6%E4%BB%96%E9%87%8F%E5%8C%96%E7%A8%8B%E5%BA%8F%E4%BC%A0%E8%BD%A8%E9%81%93.md#7-pyscf%E4%BC%A0%E8%BD%A8%E9%81%93%E7%BB%99orca) API and `mkl2py` utility may do some help.

**Comment by verena-neufeld** [comment]
Thanks for your question @anmorgunov . Has it been resolved and can this issue be closed?

**Comment by anmorgunov** [comment]
yes, it was a helpful thread and my issue is resolved, thanks!
